### DISCUSSION TITLE
Correct way to proxy a websocket connection?

### DISCUSSION THREAD

**Discussion by prescience-data** [uNetworking/uWebSockets.js#510] [Q&A]
I'm currently using `ws` to proxy a websocket connection that needs to have each message mutated as they pass through. 

Looking at the performance difference it seems crazy not to use `uws`, but I'm trying to figure out what the equivalent of `import { Server } from "ws"` is?

This is current `ws` code if it helps?

```typescript

import WebSocket, { Data, Server } from "ws"

interface Sockets {
  internal: WebSocket
  external: WebSocket | undefined
}

// Define server address.
const port: number = 80
const host: string = "192.168.1.10"

// Create socket context.
const sockets: Sockets = {
  internal: new WebSocket("ws://192.168.1.20:3000"),
  external: undefined
}

// Initialise server.
const server: Server = new Server({ port, host })

// Define close hook.
const socketClosed = (sockets: Sockets) => async (): Promise<void> => {
  await Promise.all([sockets.external?.close(), sockets.internal?.close()])
}

// Define inbound message mutation hook.
const inboundMessage = (external: WebSocket) => async (
  data: Data
): Promise<void> => {
  // ... Mutate the socket data...
  external.send(data)
}

// Define outbound message mutation hook.
const outboundMessage = (internal: WebSocket) => async (
  data: Data
): Promise<void> => {
  // ... Mutate the socket data...
  internal.send(data)
}

// Proxy sockets upon external connection.
server.on("connection", (ws: WebSocket): void => {
  sockets.external = ws

  sockets.external.on("message", inboundMessage(sockets.internal))
  sockets.external.on("close", socketClosed(sockets))

  sockets.internal.on("message", outboundMessage(sockets.external))
  sockets.external.on("close", socketClosed(sockets))
})

```

**Comment by dalisoft** [comment]
@prescience-data The example you provided is on `ws` and you saying about `uws` which is deprecated (maybe i'm wrong), you should migrate to uWebSockets.js.
uWebSockets.js replaces not only `ws` part, a entire REST/Application part and you get huge boost in performance

**Reply by prescience-data** [reply]
I was using `uws` as a shorthand for `uwebsockets.js`, didn't realise they were separated. 
The example is my current `ws` code and I'm trying for work out how to migrate it to `uwebsockets.js`

**Reply by dalisoft** [reply]
@prescience-data You can try  [this](https://github.com/nanoexpress/nanoexpress/blob/master/examples/websocket.js)
